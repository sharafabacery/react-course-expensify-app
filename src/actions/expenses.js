
import database from '../firebase/firebase'
//ADD_EXPENSE 
export const addExpense=(expense)=>({
    
    type:'ADD_EXPENSE',
    expense
})
export const startAddExpense=(expenseDate={})=>{
    return (dispatch,getState)=>{
        const uid=getState().auth.uid.uid
       
        const {description='',note='',amount=0,createdAt=0 }=expenseDate
        const expense={description,note,amount,createdAt}
        return database.ref(`users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    }
}


//REMOVE_EXPENSE
export const removeExpense=({
id
})=>({
  type:'REMOVE_EXPENSE'
  ,id 
})

//EDIT_EXPENSE

export const editExpense=(id,updates)=>({
    type:'EDIT_EXPENSE',
    id,
    updates
})

export const setExpenses=(expenses)=>({
   type:'SET_EXPENSES',
    expenses
});

export const startSetExpenses=()=>{
   return(dispatch,getState)=>{
    const uid=getState().auth.uid.uid
     return database.ref(`users/${uid}/expenses`)
        .once('value')
        .then((snapshot)=>{
            const expensesArray=[]
            snapshot.forEach((childsnapshot)=>{
                expensesArray.push({
                    id:childsnapshot.key,
                    ...childsnapshot.val()
                })
            })
           dispatch(setExpenses(expensesArray))
        }
        )




   }

};
export const startEditExpense=(id,updates)=>{
    return(dispatch,getState)=>{
        const uid=getState().auth.uid.uid
        return database.ref(`users/${uid}/expenses/${id}`).update(updates)
        .then((snapshot)=>{
           dispatch(editExpense(id,updates))
        })
    }
}
export const startremoveExpense=({id})=>{
   return(dispatch,getState)=>{
    const uid=getState().auth.uid.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then(()=>{

        dispatch(removeExpense({id}))
      })
     


   }
  



}