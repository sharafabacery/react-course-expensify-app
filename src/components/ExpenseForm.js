import React,{useState} from "react";
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'


const ExpenseForm=(props)=>{
    const[description,setDescription]=useState(props.expense?props.expense.description:'')
    const[note,setNote]=useState(props.expense?props.expense.note:'')
    const[amount,setAmount]=useState(props.expense?props.expense.amount.toString():'')
    const[createdAt,setCreatedAt]=useState(props.expense?moment( props.expense.createdAt):moment())
    const[calendarFocused,setCalendarFocused]=useState(false)
    const [error,setError]=useState('')
    const onDescriptionChange=(e)=>{
        const description=e.target.value
        setDescription(description)
        }
    const onNoteChange=(e)=>{
        const note=e.target.value
        setNote(note)
        }

    const onAmountChange=(e)=>{
        const amount=e.target.value
        if (!amount||amount.match(/^\d{0,}(\.\d{0,2})?$/)) {
                setAmount(amount)
            }
        }
    const  onDateChange=(createdAt)=>{
        if(createdAt)
        setCreatedAt(createdAt)
        }
    const  onFocusChange=({focused})=>{
                setCalendarFocused(focused)
        }
    const   onSubmit=(e)=>{
                e.preventDefault()
                if (!description || !amount) {
                    setError('Please provide description and amount')
                   
                }else{
                    setError('')
                    props.onSubmit({
                        description:description,
                        amount:parseFloat(amount,10),
                        createdAt: createdAt.valueOf(),
                        note:note
                    })
                }
        
            }

            return(
                <div>
                <p>{error!=='' &&error}</p>
                <form onSubmit={onSubmit}>
                <input type="text" placeholder='Description' autoFocus value={description} onChange={onDescriptionChange} />
                <input type="number" placeholder='Amount' value={amount} onChange={onAmountChange}/>
                    <SingleDatePicker
                        date={createdAt}
                        onDateChange={onDateChange}
                        focused={calendarFocused}
                        onFocusChange={onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={()=>false}
                       
                    />
                <textarea placeholder="Add Note for Expense (optional)" value={note} onChange={onNoteChange}></textarea>
                
                <button type="submit">Add Expense</button>
                </form>
                </div>
                )


}


export default ExpenseForm