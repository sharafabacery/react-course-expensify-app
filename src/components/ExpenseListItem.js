import React from 'react'
import {Link} from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'
export const ExpenseListItem=(props)=>(
    <div>
       <Link to={'/edit/'+props.expense.id} ><p>{props.expense.description}</p></Link>
    <p>
    {numeral(props.expense.amount).format('$0,0.00')}
     - 
     {moment(props.expense.createdAt).format('MMMM Do, YYYY')}
     </p>
    
    </div>
)

export default  ExpenseListItem