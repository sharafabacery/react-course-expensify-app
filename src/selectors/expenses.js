import moment from 'moment';

// Get visible expenses
 const expenses=(expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    return startDateMatch && endDateMatch && textMatch;
  })
  .sort((a, b) =>sortBy==='date' ? a.createdAt < b.createdAt ? 1 : -1: a.amount < b.amount ? 1 : -1);
};
export default expenses
