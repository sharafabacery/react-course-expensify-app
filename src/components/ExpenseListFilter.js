import React,{useState} from 'react'
import {connect} from 'react-redux'
import { DateRangePicker } from 'react-dates';
import {setTextFilter,sortByAmount,sortByDate,setEndDate,setStartDate  } from '../actions/filters';


export const ExpenseListFilter=(props)=>{
    const[calendarFocused,setCalendarFocused]=useState(null);

   const onDatesChange=({startDate,endDate})=>{
    props.setStartDate(startDate)
    props.setEndDate(endDate)
    }

   const onFocusChange=(calendarFocused)=>{
       setCalendarFocused(calendarFocused)

    }
   const onSortChange=(e)=>{
        if ( e.target.value==='date') {
            props.sortByDate()
        }else if(e.target.value ==='amount'){props.sortByAmount()}
       
        
            }
   const onTextChange=(e)=>{props.setTextFilter(e.target.value)}
   
            return(<div>
            <input type="text" value={props.filters.text} onChange={
               onTextChange
              
                
            }/>
            <select value={props.filters.sortBy} onChange={onSortChange}>
            <option value='date'>Date</option>
            <option value='amount'>Amount</option>
            </select>
            <DateRangePicker
            startDate={props.filters.startDate}
            endDate={props.filters.endDate}
            onDatesChange={onDatesChange}
            focusedInput={calendarFocused}
            onFocusChange={onFocusChange}
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={()=>false}
            startDateId="start"
            endDateId="end"
            
            />
            </div>)


}

const mapStateToProps=(state)=>{
    return{
        filters:state.filter
    }
}
const mapDispatchToProps=(dispatch)=>({
    setTextFilter:(text)=>dispatch(setTextFilter(text)),
    sortByDate:()=>dispatch(sortByDate()),
    sortByAmount:()=>dispatch(sortByAmount()),
    setStartDate:(StartDate)=>dispatch(setStartDate(StartDate)),
    setEndDate:(endDate)=>dispatch(setEndDate(endDate))
})
export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter)