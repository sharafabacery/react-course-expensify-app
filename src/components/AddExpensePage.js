import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startAddExpense} from '../actions/expenses'

const AddExpensePage=(props)=>{
  const  onSubmit=(expense)=>{
        props.startAddExpense(expense)
        props.history.push('/')
     }
    
         return(<div>
     <h1>
     Add Expense
     </h1>
     <ExpenseForm 
      onSubmit={onSubmit}
     />
    
    </div>)

}
// export class AddExpensePage extends React.Component{
//     onSubmit=(expense)=>{
//         this.props.addExpense(expense)
//         this.props.history.push('/')
//      }
//      render(){
//          return(<div>
//      <h1>
//      Add Expense
//      </h1>
//      <ExpenseForm 
//       onSubmit={this.onSubmit}
//      />
    
//     </div>)
//      }

// }

const mapDispatchToProps=(dispatch)=>(
      {
        startAddExpense:(expense)=>dispatch(startAddExpense(expense))
      }
)
export default connect(undefined,mapDispatchToProps)(AddExpensePage) 