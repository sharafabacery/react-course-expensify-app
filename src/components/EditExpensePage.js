import React from 'react'
import {connect}from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {editExpense,removeExpense }from '../actions/expenses'

export const EditPage=(props)=>{
   const onSubmit=(expense)=>{
    props.editExpense(props.expense.id,expense)
    props.history.push('/')
    }

   const onRemove=()=>{
    props.removeExpense({id:props.expense.id})
    props.history.push('/')
   }
   return(
        <div>
        <ExpenseForm
        expense={props.expense}
        onSubmit={onSubmit}
        />
        <button onClick={onRemove}>Remove </button>
        </div>)


}

const mapStatetoProps=(state,props)=>{
     return{
         expense:state.expenses.find((expense)=>
                expense.id === props.match.params.id
         )
     }
}

const mapDispatchToProps=(dispatch)=>({
    editExpense:(id,expense)=>dispatch(editExpense(id,expense)),
    removeExpense:(data)=>dispatch(removeExpense(data))
})
export default connect(mapStatetoProps,mapDispatchToProps)( EditPage)