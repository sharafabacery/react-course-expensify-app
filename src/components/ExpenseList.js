import React from "react";
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from "../selectors/expenses";
export const ExpenseList=(props)=>(
   
<div>
<h1>Expense List</h1>
{
   props.expenses.length ===0?(<p>No Expenses</p>):(props.expenses.map((expense,index)=>(<ExpenseListItem key={expense.id} expense={expense}/>)))
}

</div>
)

// const ConnectedExpenseList=connect((state)=>{
//     return{
//         expenses:state.expenses
//     }
// })(ExpenseList)
//map store state to compoent state
const mapStateToProps=(state)=>{
         return{
            expenses:selectExpenses(state.expenses,state.filter)
         }
     }
export default connect(mapStateToProps)(ExpenseList)