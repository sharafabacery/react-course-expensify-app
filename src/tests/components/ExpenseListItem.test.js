import React from 'react'
import {shallow} from 'enzyme'
//import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expeneses'
import {ExpenseListItem} from '../../components/ExpenseListItem';

test('should render ExpenseListItem with expense',()=>{
    const wrapper=shallow(<ExpenseListItem {...expenses[2]}/>)
    expect(wrapper).toMatchSnapshot()
})
// test('should render ExpenseListItem with empty expense',()=>{
//     const wrapper=shallow(<ExpenseListItem expense={[]}/>)
//     expect(wrapper).toMatchSnapshot()
// })