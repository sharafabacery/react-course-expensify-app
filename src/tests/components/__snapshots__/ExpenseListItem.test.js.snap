// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <Link
    to="/edit/3"
  >
    <p>
      Credit Card
    </p>
  </Link>
  <p>
    4500
     - 
    345600000
  </p>
</div>
`;
